<a href="https://colab.research.google.com/github/22Ifeoma22/22Ifeoma22/blob/main/NLP_Lesson_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1><div align="center">Natural Language Processing From Scratch</div></h1>
<div align="center">Bruno Gonçalves</div>
<div align="center"><a href="http://www.data4sci.com/">www.data4sci.com</a></div>
<div align="center">@bgoncalves, @data4sci</div>

# Lesson 3 - Sentiment Analysis

In [ ]:
import string
import gzip
from collections import Counter
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.cm as cm

%matplotlib inline

# Word counting

We start by taking the simplest approach and simply counting positive and negative words. We'll use Hu and Liu's Lexicon from their 2004 KDD paper: https://www.cs.uic.edu/~liub/FBS/sentiment-analysis.html

In [ ]:
pos = np.loadtxt('data/positive-words.txt', dtype='str', comments=';')
neg = np.loadtxt('data/negative-words.txt', dtype='str', comments=';')

Create a dictionary and assign the valence to each positive and negative word

In [ ]:
valence = {}

for word in pos:
    valence[word.lower()] = 1

for word in neg:
    valence[word.lower()] = -1

Here's the simple word extraction function we defined in Lesson I

In [ ]:
def extract_words(text):
    temp = text.split() # Split the text on whitespace
    text_words = []

    for word in temp:
        # Remove any punctuation characters present in the beginning of the word
        while word[0] in string.punctuation:
            word = word[1:]

        # Remove any punctuation characters present in the end of the word
        while word[-1] in string.punctuation:
            word = word[:-1]

        # Append this word into our list of words.
        text_words.append(word.lower())

    return text_words

That now we can use to define a sentiment measuring function that returns the valence of a sentence or piece of text. Notice that we use the valence directly from the dictionary instead of treating positive and negative words separatly. This will prove useful later on ;)

In [ ]:
def sentiment(text, valence):
    words = extract_words(text.lower())

    word_count = 0
    score = 0

    for word in words:
        if word in valence:
            score += valence[word]
            word_count += 1

    return score/word_count

Now let's test our simple code with some simple examples

In [ ]:
texts = ["I'm very happy",
         "The product is pretty annoying, and I hate it",
         "I'm sad",
        ]

for text in texts:
    print(sentiment(text, valence))

1.0
-0.3333333333333333
-1.0


This is a bit surprising. One might expect the second sentence to be negative, after all "pretty annoying" and "hate" sound pretty negative. However, since each word in taken by itself, regardless of context we end up with:

In [ ]:
words = extract_words(texts[1].lower())
for word in words:
    if word in valence:
        print(word, valence[word])

pretty 1
annoying -1
hate -1


We'll see in a bit how to handle cases like this, but the solution requires two important changes to our current approach: modifier words and real valued weights

# Modifiers

The first step is to define a dictionary of modifiers

In [ ]:
modifiers = {
    "very": 1.5,
    "much": 1.3,
    "not": -1,
    "pretty": 1.5,
    "somewhat": 1.2}

And to change our sentiment measuring function to take the modifiers into account.

In [ ]:
def sentiment_modified(text, valence, modifiers, verbose=False):
    words = extract_words(text.lower())

    word_count = 0
    score = 0
    ngrams = [[]]

    # generate ngrams
    for i in range(len(words)):
        word = words[i]

        if word in modifiers:
            ngrams[-1].append(word)
            continue

        if word in valence:
            ngrams[-1].append(word)
        else:
            if len(ngrams[-1]) > 0:
                ngrams.append([])

    score = 0

    # Remove the trailing empty ngram if necessary
    if len(ngrams[-1]) == 0:
        ngrams = ngrams[:-1]

    for ngram in ngrams:
        value = 1

        for word in ngram:
            if word in modifiers:
                value *= modifiers[word]
            elif word in valence:
                value *= valence[word]

        if verbose:
            print(ngram, value)

        score += value

    return score/len(ngrams)

This implementation is still relatively simple, but, as you can see, the results are already better.

In [ ]:
sentiment_modified(texts[1], valence, modifiers, True)

['pretty', 'annoying'] -1.5
['hate'] -1


-1.25

A more complete implementation would be more careful in handling the modifiers and would build larger ngrams so that cases like this one would also work:

In [ ]:
sentiment_modified("It was not very good", valence, modifiers, True)

['not', 'very', 'good'] -1.5


-1.5

And even more complex (and unrealistic) examples work fine

In [ ]:
sentiment_modified("It was not not very very good", valence, modifiers, True)

['not', 'not', 'very', 'very', 'good'] 2.25


2.25

# Continuous weights

VADER is a state of the art sentiment analysis tool. Here we will use their excelent and well documented [lexicon](https://github.com/cjhutto/vaderSentiment) to explore non binary weights. Their approach is significantly more advanced than what we present here, but some of the fundamental ideas are the same

In [ ]:
vader = pd.read_csv("data/vader_lexicon.txt", sep='\t', header=None)

The vader lexicon includes a lot of interesting information:

In [ ]:
vader.head()

0    1        2                                         3
0   $: -1.5  0.80623  [-1, -1, -1, -1, -3, -1, -3, -1, -2, -1]
1   %) -0.4  1.01980       [-1, 0, -1, 0, 0, -2, -1, 2, -1, 0]
2  %-) -1.5  1.43178    [-2, 0, -2, -2, -1, 2, -2, -3, -2, -3]
3  &-: -0.4  1.42829      [-3, -1, 0, 0, -1, -1, -1, 2, -1, 2]
4   &: -0.7  0.64031    [0, -1, -1, -1, 1, -1, -1, -1, -1, -1]

Similies are also included and, in addition to the average sentiment of each word (in column 1) and it's standard deviation (in column 2) it provides the raw human generated scores in column 3. So that we may easily check (and possibly modify) their weights. To extract the raw scores for the word "love" we could simply do:

In [ ]:
scores = eval(vader.iloc[4446][3])
print(scores)

[3, 3, 3, 3, 3, 3, 3, 4, 4, 3]


And we can see that 8/10 people thought that the word love should receive a score of 3 and two others a score of 4. This gives us insight into how uniform the scores are.  If for some reason, we thought that there was some problem with the 2 values of 4 or perhaps just not appropriate to our purposes we might discard them and recalculate the valence of the word.

One justification for this might be the fact that the scores for the closely related word, "loved", are significantly different with a wider range of variation in the human scores

In [ ]:
vader.iloc[4447]

0                             loved
1                               2.9
2                               0.7
3    [3, 3, 4, 2, 2, 4, 3, 2, 3, 3]
Name: 4447, dtype: object

Now we convert this dataset into a dictionary similar to the one we used above

In [ ]:
valence_vader = dict(vader[[0,1]].values)

In [ ]:
valence_vader[':)']

2.0

To use this new dictionary we just have to modify the arguments to the sentiment_modified function:

In [ ]:
sentiment_modified("It was not not very very good", valence_vader, modifiers, True)

['not', 'not', 'very', 'very', 'good'] 4.2749999999999995


4.2749999999999995

One important detail to keep in mind is that scores obtained through different methods are not comparable. In this example, the score of the sentence "It was not not very very good" went from 2.25 to 4.27 when we switched dictionaries. This is due not only to different levels of coverage in differnet dictionaries but also to differnet choices in the possible ranges of values.